crates/red_knot/src/main.rs

```diff
@@ -48,225 +44,305 @@ fn main() -> anyhow::Result<()> {
-                            // avoid stale messages in the next main loop.
-                            run_cancellation_token_source.cancel();
+impl Drop for MainLoop {
+    fn drop(&mut self) {
+        self.orchestrator_sender
+            .send(OrchestratorMessage::Shutdown)
+            .unwrap();
+    }
+}
 
-                            aggregated_changes.extend(changes);
+#[derive(Debug, Clone)]
+struct FileChangesNotifier {
+    sender: crossbeam_channel::Sender<OrchestratorMessage>,
+}
 
-                            if pending_analysis.load(Ordering::SeqCst) == 0 {
-                                return consume_pending_messages(&receiver, aggregated_changes);
-                            }
-                        }
-                    }
+impl FileChangesNotifier {
+    fn notify(&self, changes: Vec<FileChange>) {
+        self.sender
+            .send(OrchestratorMessage::FileChanges(changes))
+            .unwrap();
+    }
+}
+
+#[derive(Debug)]
+struct MainLoopCancellationToken {
+    sender: crossbeam_channel::Sender<MainLoopMessage>,
+}
+
+impl MainLoopCancellationToken {
+    fn stop(self) {
+        self.sender.send(MainLoopMessage::Exit).unwrap();
+    }
+}
+
+struct Orchestrator {
+    aggregated_changes: AggregatedChanges,
+    pending_analysis: Option<PendingAnalysisState>,
+
+    /// Sends messages to the main loop.
+    sender: crossbeam_channel::Sender<MainLoopMessage>,
+    /// Receives messages from the main loop.
+    receiver: crossbeam_channel::Receiver<OrchestratorMessage>,
+}
+
+impl Orchestrator {
+    fn run(&mut self) {
+        while let Ok(message) = self.receiver.recv() {
+            match message {
+                OrchestratorMessage::Run => {
+                    self.pending_analysis = None;
+                    self.sender.send(MainLoopMessage::CheckProgram).unwrap();
                 }
 
-                // This can be reached if there's no Ctrl+C and no file watcher handler.
-                // In that case, assume that we don't run in watch mode and exit.
-                NextTickCommand::Exit
-            });
-
-            // Star the analysis task on the thread pool and wait until they complete.
-            rayon::scope(|scope| {
-                for file in &files_to_check {
-                    let cancellation_token = run_cancellation_token.clone();
-                    if cancellation_token.is_cancelled() {
-                        break;
-                    }
+                OrchestratorMessage::CheckProgramStarted { cancellation_token } => {
+                    debug_assert!(self.pending_analysis.is_none());
 
-                    let sender = sender.clone();
+                    self.pending_analysis = Some(PendingAnalysisState { cancellation_token });
+                }
 
-                    started_analysis.fetch_add(1, Ordering::SeqCst);
+                OrchestratorMessage::CheckProgramCompleted(diagnostics) => {
+                    self.pending_analysis
```

Comment:
I consider it a bug if the state machine are out of sync and would prefer if we panic over trying to make it work somehow (and leak memory).

Reply:
That's reasonable. I wasn't suggesting we try to recover, more just that this could help clarify what happened in the bug case. But maybe tracing is sufficient for that.